### PR TITLE
✨clusterctl upgrade apply: wait for providers to be ready

### LINKED ISSUE #6267: add flags to `clusterctl upgrade apply` to wait for the upgrade to finish before exiting
**User Story**

A a user of `clusterctl` I would like the `clusterctl upgrade apply` command to wait for the new components to be available before the command exits. Similar to how `clusterctl init` enables this by using the `--wait-providers` and `--wait-provider-timeout` flags.
Adding flags with the same names ,`--wait-providers` and `--wait-provider-timeout`, should be fine.

**Detailed Description**

Currently after performing `clusterctl upgrade apply` the command exists immediately after the components of the new providers are installed. It would be nice to have the feature to wait for the new deployments installed using a `--wait-providers` flag (similar to init). This feature will also be useful in identifying if any of the new deployments are not spinning up as expected.

*Additional note: After this feature is implemented there we can probably refactor some our e2e tests to remove the wait logic and instead just call the upgrade command with the corresponding wait flags*

**Anything else you would like to add:**

We already have helper functions in place that are used as part of the `init` command that can be repurposed (with a little refactor) to work with the `upgrade apply` command as well.
https://github.com/kubernetes-sigs/cluster-api/blob/5026786ee809c5def466049f5befd2a786fbcefa/cmd/clusterctl/client/cluster/installer.go#L117-L129

Adding a similar wait logic after the for loop at end (after the deleting the webhook namespace section) when all the providers are updated might be a good place to start
https://github.com/kubernetes-sigs/cluster-api/blob/fa6bf66a0f12c57d3ebfb362fd634db6c62eabc6/cmd/clusterctl/client/cluster/upgrader.go#L383-L410

/kind feature

### PR DESCRIPTION
**Pull request by praveenrewar**
**What this PR does / why we need it**:
`clusterctl upgrade apply` should wait for the cluster to be 'ready' before exiting.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6267

### CODE REVIEW

File: cmd/clusterctl/cmd/upgrade_apply.go [hunk 1]
```diff
@@ -73,6 +77,10 @@ func init() {
 		"Bootstrap providers instance and versions (e.g. capi-kubeadm-bootstrap-system/kubeadm:v0.3.0) to upgrade to. This flag can be used as alternative to --contract.")
 	upgradeApplyCmd.Flags().StringSliceVarP(&ua.controlPlaneProviders, "control-plane", "c", nil,
 		"ControlPlane providers instance and versions (e.g. capi-kubeadm-control-plane-system/kubeadm:v0.3.0) to upgrade to. This flag can be used as alternative to --contract.")
+	upgradeApplyCmd.Flags().BoolVar(&ua.waitProviders, "wait-providers", false,
+		"Wait for providers to be installed.")
```
Comment (ykakarap) [inline cmd/clusterctl/cmd/upgrade_apply.go:81]:
```suggestion
		"Wait for providers to be upgraded.")
```

File: cmd/clusterctl/cmd/upgrade_apply.go [hunk 2]
```diff
@@ -73,6 +77,10 @@ func init() {
...
+	upgradeApplyCmd.Flags().IntVar(&ua.waitProviderTimeout, "wait-provider-timeout", 5*60,
+		"Wait timeout per provider installation in seconds. This value is ignored if --wait-providers is false")
```
Comment (ykakarap) [inline cmd/clusterctl/cmd/upgrade_apply.go:83]:
```suggestion
		"Wait timeout per provider upgrade in seconds. This value is ignored if --wait-providers is false")
```

File: cmd/clusterctl/client/upgrade.go [hunk 3]
```diff
@@ -107,6 +108,12 @@ type ApplyUpgradeOptions struct {
 
 	// InfrastructureProviders instance and versions (e.g. capa-system/aws:v0.5.0) to upgrade to. This field can be used as alternative to Contract.
 	InfrastructureProviders []string
+
+	// WaitProviders instructs the upgrade apply command to wait till the providers are installed.
```
Comment (ykakarap) [inline cmd/clusterctl/client/upgrade.go:112]:
```suggestion
	// WaitProviders instructs the upgrade apply command to wait till the providers are successfully upgraded.
```

File: cmd/clusterctl/client/cluster/upgrader.go [hunk 4]
```diff
@@ -44,10 +44,10 @@ type ProviderUpgrader interface {
 	Plan() ([]UpgradePlan, error)
 
 	// ApplyPlan executes an upgrade following an UpgradePlan generated by clusterctl.
-	ApplyPlan(clusterAPIVersion string) error
+	ApplyPlan(opts InstallOptions, clusterAPIVersion string) error
 
 	// ApplyCustomPlan plan executes an upgrade using the UpgradeItems provided by the user.
-	ApplyCustomPlan(providersToUpgrade ...UpgradeItem) error
+	ApplyCustomPlan(opts InstallOptions, providersToUpgrade ...UpgradeItem) error
```
Comment (ykakarap) [inline cmd/clusterctl/client/cluster/upgrader.go:50]: These are breaking API changes. The last time we introduced similar flags to `clusterctl init` we made some breaking changes citing that "We are not currently providing any compatibility guarantees on cluserctl used as a library". Ref: https://github.com/kubernetes-sigs/cluster-api/pull/4934/files#r683272400

@fabriziopandini Can you please confirm if we can still go ahead and make breaking change to the clusterctl library API?

**Review by ykakarap — COMMENTED** [review]
@praveenrewar  Thank you for working on this! 🚀 

I am LGTM with the overall approach. I will take another look after we have more information about the breaking API change.

**Review by ykakarap — COMMENTED** [review]
/ok-to-test

File: cmd/clusterctl/client/cluster/upgrader.go [hunk 5]
```diff
@@ -44,10 +44,10 @@ type ProviderUpgrader interface {
 	Plan() ([]UpgradePlan, error)
 
 	// ApplyPlan executes an upgrade following an UpgradePlan generated by clusterctl.
-	ApplyPlan(clusterAPIVersion string) error
+	ApplyPlan(opts InstallOptions, clusterAPIVersion string) error
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/upgrader.go:47]: Should this here and below be UpgradeOptions? I think we should not re-use the options struct from the ProviderInstaller

Comment (praveenrewar) [reply cmd/clusterctl/client/cluster/upgrader.go:47]: Yep, makes sense. I will make the change.

Comment (praveenrewar) [reply cmd/clusterctl/client/cluster/upgrader.go]: Done.

File: cmd/clusterctl/cmd/upgrade_apply.go (hunk 1 above)
Comment (sbueringer) [inline cmd/clusterctl/cmd/upgrade_apply.go:81]:
```suggestion
		"Wait for providers to be upgraded.")
```

File: cmd/clusterctl/cmd/upgrade_apply.go (hunk 2 above)
Comment (sbueringer) [inline cmd/clusterctl/cmd/upgrade_apply.go:83]:
```suggestion
		"Wait timeout per provider upgrade in seconds. This value is ignored if --wait-providers is false")
```

File: cmd/clusterctl/client/upgrade.go (hunk 3 above)
Comment (sbueringer) [inline cmd/clusterctl/client/upgrade.go:112]:
```suggestion
	// WaitProviders instructs the upgrade apply command to wait till the providers are successfully upgraded.
```

File: cmd/clusterctl/client/upgrade.go [hunk 6]
```diff
@@ -107,6 +108,12 @@ type ApplyUpgradeOptions struct {
...
+	WaitProviders bool
+
+	// WaitProviderTimeout sets the timeout per provider wait installation
```
Comment (sbueringer) [inline cmd/clusterctl/client/upgrade.go:115]:
```suggestion
	// WaitProviderTimeout sets the timeout per provider upgrade.
```

**Review by sbueringer — COMMENTED** [review]
Just a few nits.

P.S. Yuvaraj commented on them before as well, did you maybe only forget to push a commit?

**Comment by praveenrewar** [comment]
> P.S. Yuvaraj commented on them before as well, did you maybe only forget to push a commit?

Oops, my bad. Good catch! Thanks :)

**Comment by praveenrewar** [comment]
From the issue:
> Additional note: After this feature is implemented there we can probably refactor some our e2e tests to remove the wait logic and instead just call the upgrade command with the corresponding wait flags

Should I update the e2e tests as part of the same PR or does it make sense to have a separate PR for it?
(I am leaning towards making the change in this PR itself but okay with both the options)

**Comment by sbueringer** [comment]
> Oops, my bad. Good catch! Thanks :)

Thx for the quick fix :). 

> Should I update the e2e tests as part of the same PR or does it make sense to have a separate PR for it? (I am leaning towards making the change in this PR itself but okay with both the options)

I would suggest to make it a separate PR. I think this PR now would be ready for a final review from someone else + merge.


Can you please squash the commits? (we're running Prow in merge mode for our release notes tool, so we have to merge manually on the PRs ourselves)

**Comment by praveenrewar** [comment]
> I would suggest to make it a separate PR. I think this PR now would be ready for a final review from someone else + merge.

Cool 👍🏻 

> Can you please squash the commits?

Sure, done :)

**Comment by sbueringer** [comment]
Thx!

/lgtm
/assign @ykakarap

**Comment by praveenrewar** [comment]
Hey @ykakarap, do let me know if anything else is required on this PR :)

**Comment by sbueringer** [comment]
/assign @fabriziopandini 
for a final look

**Review by ykakarap — APPROVED** [review]
/lgtm

>Should I update the e2e tests as part of the same PR or does it make sense to have a separate PR for it?
>(I am leaning towards making the change in this PR itself but okay with both the options)

A separate PR for that would be ideal.

**Pull request merged**